### ISSUE TITLE
perceptron ML7

### ISSUE THREAD

**Issue by henryliangt** [henryliangt/usyd#32]
![image](https://user-images.githubusercontent.com/23263731/200296378-68401e7c-8b40-4349-a4b7-9c7533b96fe8.png)

![image](https://user-images.githubusercontent.com/23263731/200296437-923e044e-81b3-406e-9911-b0348903f077.png)

![image](https://user-images.githubusercontent.com/23263731/200297021-92ab28f1-ae4a-40eb-b8c1-062fb9f9c425.png)

**Comment by henryliangt** [comment]
Perceptron learning rule

![image](https://user-images.githubusercontent.com/23263731/200297098-ffe3f2e7-71e0-4889-b46d-2a3185ad49d5.png)

Perceptron learning algorithm
![image](https://user-images.githubusercontent.com/23263731/200297214-3c29e783-7ef9-44e1-b83b-0a7661adb3e3.png)

Checking the stopping condition

![image](https://user-images.githubusercontent.com/23263731/200297298-4cb991dc-d843-4da3-8f32-c771523b97b5.png)

**Comment by henryliangt** [comment]
Perceptron - example
![image](https://user-images.githubusercontent.com/23263731/200297397-86cb54ca-c749-4ad4-8de9-e1d1ad9d9a3d.png)

Solution (1)
![image](https://user-images.githubusercontent.com/23263731/200297475-b5a79561-74f0-43f6-bc3f-8b56d0492d1f.png)

Solution (2)
![image](https://user-images.githubusercontent.com/23263731/200297550-c9b96cbe-7091-4810-a987-9ee068b57a86.png)

Perceptron - capabilities
![image](https://user-images.githubusercontent.com/23263731/200297623-da01ac38-a15c-43fa-a7f2-c7a92d9faed0.png)


Perceptron - history

![image](https://user-images.githubusercontent.com/23263731/200297685-1e32dd21-ef85-4923-ba06-a88d7fe7eae1.png)

**Comment by henryliangt** [comment]
Perceptron for AND gate

![image](https://user-images.githubusercontent.com/23263731/200297806-fea76c1b-6ee6-4fcb-a44e-eaf90f7e9060.png)

Perceptron for OR gate
![image](https://user-images.githubusercontent.com/23263731/200297886-c549f41c-8c5d-421d-ba44-d4276eb2c17f.png)

Perceptron for NAND gate
![image](https://user-images.githubusercontent.com/23263731/200297933-5c37dc3a-14f9-4245-a2f9-3311afad1b64.png)

Perceptron for XOR gate

![image](https://user-images.githubusercontent.com/23263731/200297985-4545df45-6306-4964-a681-60f10f3a4a7c.png)

**Comment by henryliangt** [comment]
Perceptron for XOR gate 2
![image](https://user-images.githubusercontent.com/23263731/200298064-c310df3c-4692-4100-9268-ae15a5f42fb4.png)

Adding more layers
![image](https://user-images.githubusercontent.com/23263731/200298131-7fa60821-f9d2-48ec-89f3-2650f35775a2.png)

**Comment by henryliangt** [comment]
Multi-layer perceptrons and backpropagation algorithm
**Backpropagation algorithm**
• An algorithm that is used to train multi-layer perceptron neural 
networks
• Proposed by Paul Werbos in 1974; later re-discovered by David 
Rumelhart, Geoffrey Hinton, Ronald Williams 1986; David Parker 
1985; Yann LeCun 1985

**Network architecture**
• Layers of neurons – input layer, output layer, 1 or mode hidden layers
• Feedforward NN – each neuron receives input only from the previous layer
• Fully connected network – each neuron in the current layer is connected with
all neurons from the previous

![image](https://user-images.githubusercontent.com/23263731/200298412-876cb0a4-3cbe-4d18-8c80-198214c9719f.png)


**Neuron**
• A single neuron calculates the weighed sum of the outputs of the previous 
layer, which is passed through a transfer function
• The transfer function needs to be differentiable
• Most widely used transfer function: sigmoid
![image](https://user-images.githubusercontent.com/23263731/200298501-e4bd8986-3132-4916-abcf-8a8a93aa6e8a.png)


**Backpropagation algorithm - idea**

• For each training example {x,t}, x={x1
, x2
, ..xn
}
• Propagate x through the network and calculate the output o. Compare it with 
the target output t and calculate the error.
• Update weights of the network to reduce the error
• Until error over all examples < threshold
• Adjusts the weights backwards - from the output to the input neurons by 
propagating the weight change to minimize the error

![image](https://user-images.githubusercontent.com/23263731/200298622-cb88e4bd-fe4f-4f0e-845d-3b6dbaca53d8.png)


• How to calculate the weight change?
• By defining an error function and using the gradient descent algorithm to 
calculate it


**Steepest gradient descent** 
• We define an error function (also called cost or loss function), e.g. MSE over all 
training examples
• We can minimize it by using the steepest gradient descent algorithm (standard 
optimization method)
• The NN weights are iteratively updated by moving downhill - in the direction that 
reduces the error the most – this is the direction of the negative of the gradient
![image](https://user-images.githubusercontent.com/23263731/200298743-dc6f4856-da34-4226-8314-42d6ce233759.png)

**Comment by henryliangt** [comment]
**Gradient descent (2)**
![image](https://user-images.githubusercontent.com/23263731/200298853-ab826417-008d-4d86-9cfb-12bff3f44b6c.png)

**Local and global minimum**

 
![image](https://user-images.githubusercontent.com/23263731/200298904-695f26a6-3a6d-4803-862b-15366b6705a0.png)

**Weight update formulas** 
![image](https://user-images.githubusercontent.com/23263731/200298951-cfc56322-c49e-4245-9e01-7b561c785c60.png)

**Weight update for sigmoid transfer function**
![image](https://user-images.githubusercontent.com/23263731/200299009-c595d9ec-933d-4029-9a90-eee4fabc1673.png)

**Comment by henryliangt** [comment]
**Training NN with the backpropagation algorithm**

1. Initialize all weights (biases incl.) to small random values, e.g. [-1,1] 
• 2. Repeat until stopping condition is satisfied:
• (forward pass)
• --Present a training example and compute the network output
• (backward pass) 
• --Compute the  values for the output neurons and update the 
weights to the output layer 
• --Starting with output layer, repeat for each layer in the network:
• propagate the  values back to the previous layer
• update the weights between the two layers
• 3. Check the stopping condition at the end of each epoch – e.g. the 
error on the training set is below a threshold or a maximum number of 
epochs has been reached


**Backpropagation algorithm - example**
![image](https://user-images.githubusercontent.com/23263731/200299140-c4ca72f7-5291-4e42-8fda-15c71b7c2740.png)

**Backpropagation algorithm – example (2)**
![image](https://user-images.githubusercontent.com/23263731/200299221-bd0a514b-659b-4a8a-a267-3fc0cdc3b54a.png)

**Backpropagation algorithm – example (3)**

![image](https://user-images.githubusercontent.com/23263731/200299281-32c26658-9480-4a37-a0d2-512308ac1193.png)


**Backpropagation algorithm – example (4)**
![image](https://user-images.githubusercontent.com/23263731/200299343-ce3bff80-f9bc-43bf-820c-10348bae4178.png)

**Comment by henryliangt** [comment]
![image](https://user-images.githubusercontent.com/23263731/200299467-86a2e21f-908d-4057-abbd-b0e430ae34f7.png)


![image](https://user-images.githubusercontent.com/23263731/200299500-66f39988-9575-4213-9008-4c35bfe5d32c.png)


![image](https://user-images.githubusercontent.com/23263731/200299530-8d1d0cb0-5589-46a4-80f3-b6e8cdb75375.png)

![image](https://user-images.githubusercontent.com/23263731/200299559-e8a91d00-fd03-461f-800e-599cde8f1490.png)


![image](https://user-images.githubusercontent.com/23263731/200299587-26b3414c-75e0-413b-a04f-bb7dadea6625.png)


![image](https://user-images.githubusercontent.com/23263731/200299636-ef700db7-b7c2-4cf5-b963-c2a7483d9684.png)


![image](https://user-images.githubusercontent.com/23263731/200299684-879db697-816b-442f-ab16-173307b647a0.png)

**Comment by henryliangt** [comment]
**Stochastic gradient descent**
• The standard gradient descent algorithm sums the error of all training 
examples and then updates the weights
• The Stochastic Gradient Descent (SGD) updates the weights after each 
example – this is the algorithm we used; SGD is usually faster
• Mini-batch SGD (also called batch gradient descent) – sum the error of 
mini-batches of training examples, update the weights

**Universality of multi-layer perceptrons**
• Multi-layer perceptrons trained with the backpropagation algorithm 
are universal approximators – theorems:
• Any continuous function can be approximated with arbitrary small 
error by a network with one hidden layer (Cybenko 1989, Hornik
et al. 1989):
• Any function (including discontinuous) can be approximated to 
arbitrary small error by a network with two hidden layers 
(Cybenko 1988)
• These are existence theorems – they don’t say how to choose the 
network architecture and hyperparameters

**Comment by henryliangt** [comment]
Design issues - architecture
• Number of neurons in the input layer
• Numerical attributes - 1 neuron per attribute
• Categorical attributes with k values – k neurons per attribute, 
one-hot encoding 
• One-hot encoding - represent a k-valued attribute with k binary 
attributes, only 1 of which is set to 1 (“hot”) and all the others are 0
• E.g. outlook had 3 values – sunny, overcast, rainy; one-hot 
encoding: 1 0 0 for outlook=sunny, 0 1 0 for outlook=overcast and 0 
0 1 for outlook=rainy
• Number of neurons in the output layer
• 1 neuron for binary class problem
• k neurons for k-class problem, one-hot encoding

**Design issues – architecture (2)**

![image](https://user-images.githubusercontent.com/23263731/200299922-6227a23c-b2ef-4ad5-ab4f-8ca6ff2ac973.png)

**Presentation of training examples**
• Standard approach:
• Each epoch runs through the same training examples, one by one, 
always in the same sequence
• Alternatives:
• 1. For each epoch, permutate the training examples
• 2. Present more often the examples with higher error and less often the 
examples with lower error
• 3. Present the examples not one by one, but in batches of N examples, 
summing up their individual errors and updating after each batch (minibatch)


**Learning rate**
![image](https://user-images.githubusercontent.com/23263731/200300054-6e574af0-7ea1-4199-9c36-65bb08df90e6.png)


**Learning rate - constant**

![image](https://user-images.githubusercontent.com/23263731/200300105-a647e4a9-4390-4510-a67c-442e2b70c8c3.png)

Demo from https://towardsai.net/p/machine-learning/analysis-of-learning-rate-in-gradient-descent-algorithm-using-python

**Learning rate – time-dependent**
• Alternatives: variable learning rate – time-dependent
• Start with a high learning rate, gradually decrease it – motivation:
• Big learning rate initially -> greater weight change - can reduce the number 
of training epochs and may even help to jump over some local minima 
• Later - decrease the learning rate to prevent overshooting the global 
minimum
• Different decay schedules:
• 0 =
𝑛−1
1+𝑑+𝑛
(time-based) 𝑛= 0𝑒
−𝑑𝑛 (exponential)
Irena Koprinska, irena.koprinska@sydney.edu.au COMP5318 ML&DM, week 7, 2022
n – epoch number, d - decay rate (hyperparameter);
e.g. if 0=0.3, values in the first 5 epochs:
![image](https://user-images.githubusercontent.com/23263731/200300201-a03b4433-07f3-41a3-a6e9-fcbd95f75262.png)
